### PR TITLE
BUG: fix _array2string for structured array (issue #5692)

### PR DESCRIPTION
**Pull request by skwbc**
The cause of issue #5692 is that `_array2string` (in numpy/core/arrayprint.py) doesn’t have format function for structured arrays and it uses general purpose format function to format array elements of structured arrays. 
I made `StructureFormat` class to format structured array elements. `_get_format_function` instantiates `StructureFormat` by instantiating format function classes for each field of structure recursively and merge them.

### CODE REVIEW

**Comment by charris** [comment]
@skwbc You should expand the commit message to contain your comment above. Also include `Closes #5692` in the commit message body so that the issue is closed when this PR is merged.

**Comment by skwbc** [comment]
@charris 
I have modified the commit message to include my comment above and `Closes #5692`.

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -234,27 +234,23 @@ def _boolFormatter(x):
...
+    find the right formatting function for the dtype_
+    """
+    dtype_ = data.dtype
+    if dtype_.fields is not None:
+        format_functions = []
+        for descr in dtype_.descr:
+            field_name = descr[0]
+            field_values = data[field_name]
+            if 1 < len(field_values.shape):
+                format_function = repr_format
```
Comment (charris) [inline numpy/core/arrayprint.py]: Is `repr_format` the correct choice?

**Comment by charris** [comment]
LGTM. @ahaldane Might take a look.

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (skwbc) [inline numpy/core/arrayprint.py]: Maybe, yes (low confidence).
By using `repr_format`, this updated `_array2string` should be called and it works fine.

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -234,27 +234,23 @@ def _boolFormatter(x):
...
+    """
+    find the right formatting function for the dtype_
+    """
+    dtype_ = data.dtype
+    if dtype_.fields is not None:
+        format_functions = []
+        for descr in dtype_.descr:
+            field_name = descr[0]
+            field_values = data[field_name]
+            if 1 < len(field_values.shape):
```
Comment (skwbc) [inline numpy/core/arrayprint.py]: I rewrote it.

**Comment by charris** [comment]
Thanks @skwbc . If you want to take a shot at adding standard docstrings to the `arrayprint` functions that would be awesome.

**Comment by ahaldane** [comment]
Sorry, I only just saw this.

Looks like the right idea, but I see two problems:

First, this will raise an exception if the dtype has padding bytes:
```
>>> dt = np.dtype('u1,i4', align=True)
>>> print np.zeros(3, dtype=dt)
ValueError: no field of name
```
That's easy to fix though, just do
```
        for field_name in dtype_.names:
```
instead of
```
        for descr in dtype_.descr:
            field_name = descr[0]
```
This happens because `dtype.descr` is specifically designed to satisfy the PEP3118 requirements about padding bytes, and is *only* meant to be used in that context. I almost want to add a warning about that to the docstring of `dtype.descr` so that people don't use it. I think it confuses many people.

Second, I don't think this treats subarrays quite correctly:
```
>>> print np.zeros(shape=3, dtype=[('B', 'i4', 5)])
[(array([0, 0, 0, 0, 0], dtype=int32),)
 (array([0, 0, 0, 0, 0], dtype=int32),)
 (array([0, 0, 0, 0, 0], dtype=int32),)]
```
but I think it should give (like in 1.11)
```
>>> print np.zeros(shape=3, dtype=[('B', 'i4', 5)])
[([0, 0, 0, 0, 0],) ([0, 0, 0, 0, 0],) ([0, 0, 0, 0, 0],)]
```

@skwbc if you can make a new PR with to fix those, please do. If you don't in a few days one of us will.

**Comment by mhvk** [comment]
@skwbc - also see #8172 - it would be nice if this worked for structured array scalars as well!

**Comment by skwbc** [comment]
@ahaldane I'll make a new PR in a few days.

**Comment by jakirkham** [comment]
FWIW this broken some doctests I had. So is complicating my ability to test against 1.11 and 1.12. Is there a way to fix this to keeping a trailing zero or force dropping it in 1.11?

**Comment by eric-wieser** [comment]
Perhaps write your doctests as

```
>>> some_result == np.array(...)
True
```

Instead of relying on the repr. Alternatively, only run your doctests on one version of numpy?

**Comment by jakirkham** [comment]
Is there no option to specify how I would like to format floats? It seems like this changed how formatting is handled under the surface. Would expect that the users would have some recourse. Is that not the case?

**Comment by mhvk** [comment]
@jakirkham - It is indeed quite annoying and I had similar problems adjusting the doctests for astropy; my conclusion was that it was better to mimic new numpy even on old versions, but the problem is that one used to have no control over the formatting. So, I ended up writing code such that it used an updated `array2string` -- see  https://github.com/astropy/astropy/blob/master/astropy/coordinates/representation.py#L34 (ignore the `TODO` there, but note that this presumes the structured array only contains `float`).

**Comment by jakirkham** [comment]
Thanks for the tips, @mhvk. Unfortunately my structured arrays contain a variety of types outside of `float`s. That said, I think I found a viable workaround by simply overriding the `formatter` for `float` to convert to a Python `float` and then create a `str`. It's not the prettiest thing, but it seems to do the trick. Ideally one would add this to setup and teardown functions of relevant doctests. Please see commit ( https://github.com/nanshe-org/nanshe/commit/00233bc82bc3b7d5d27d8ce8363c074b4d43be83 ) for an example.

**Comment by mhvk** [comment]
Indeed, the problem was that prior to 1.12, the structured array formatters simply ignored the numpy formats (and thus deferred to python).

**Pull request merged**